#VRML_SIM R2022b utf8
# license: Copyright Cyberbotics Ltd. Licensed for use only with Webots.
# license url: https://cyberbotics.com/webots_assets_license
# documentation url: https://www.cyberbotics.com/doc/guide/surveyor
# Thymio II mobile robot.
# template language: javascript

EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2022b/projects/robots/mobsya/thymio/protos/Thymio2MainIndexedFaceSet.proto"
EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2022b/projects/robots/mobsya/thymio/protos/Thymio2SecondaryIndexedFaceSet.proto"
EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2022b/projects/robots/mobsya/thymio/protos/Thymio2WheelIndexedFaceSet.proto"
EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2022b/projects/robots/mobsya/thymio/protos/Thymio2TireIndexedFaceSet.proto"
EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2022b/projects/robots/mobsya/thymio/protos/Thymio2DistanceSensor.proto"
EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2022b/projects/robots/mobsya/thymio/protos/Thymio2LED.proto"

PROTO Thymio2 [
  field SFVec3f    translation     0 0 0           # Is `Transform.translation`.
  field SFRotation rotation        0 0 1 0         # Is `Transform.rotation`.
  field SFString   name            "Thymio II"     # Is `Solid.name`.
  field SFString   controller      "<generic>"     # Is `Robot.controller`.
  field MFString   controllerArgs  "port=33333"    # Is `Robot.controllerArgs`.
  field SFString   window          "<generic>"     # Is `Robot.window`.
  field SFString   customData      ""              # Is `Robot.customData`.
  field SFBool     supervisor      FALSE           # Is `Robot.supervisor`.
  field SFBool     synchronization TRUE            # Is `Robot.synchronization`.
  field SFString   contactMaterial "thymio body"   # Defines the `Solid.contactMaterial` for the body.
  field SFBool     castLight       TRUE            # Defines whether OpenGL lights are cerated for each LEDs.
  field MFNode     bodySlot        []              # Extends the robot with new nodes in the body slot.
]
{
  Robot {
    translation IS translation
    rotation IS rotation
    controller IS controller
    controllerArgs IS controllerArgs
    window IS window
    customData IS customData
    supervisor IS supervisor
    synchronization IS synchronization
    name IS name
    model "Thymio II"
    children [
      Transform {
        rotation 0.57735 0.57735 0.57735 2.094395
        children [
          Shape {
            appearance DEF MAIN_APPEARANCE PBRAppearance {
              baseColor 1 1 1
              roughness 0.1
              metalness 0
            }
            geometry Thymio2MainIndexedFaceSet {
            }
          }
          Shape {
            appearance PBRAppearance {
              baseColor 0.1 0.1 0.1
              roughness 0.2
              metalness 0
            }
            geometry Thymio2SecondaryIndexedFaceSet {
            }
          }
        ]
      }
      Transform {
        translation -0.02 0 0.0533
        rotation 0 0 1 -3.141589
        children [
          Shape {
            appearance PBRAppearance {
              metalness 0
              roughness 1
              baseColorMap ImageTexture {
                url [
                  "https://raw.githubusercontent.com/cyberbotics/webots/R2022b/projects/robots/mobsya/thymio/protos/textures/thymio2_grid.png"
                ]
              }
            }
            geometry Plane {
              size 0.006 0.006
            }
          }
        ]
      }
      Transform {
        translation 0.0168 0 0.0533
        rotation 0 0 1 -3.141589
        children [
          Shape {
            appearance PBRAppearance {
              metalness 0
              roughness 0.7
              baseColorMap ImageTexture {
                url [
                  "https://raw.githubusercontent.com/cyberbotics/webots/R2022b/projects/robots/mobsya/thymio/protos/textures/thymio2_battery.png"
                ]
              }
            }
            geometry Plane {
              size 0.006 0.01
            }
          }
        ]
      }
      Transform {
        translation 0.0168 0 0.0533
        rotation 0 0 1 -3.141589
        children [
          Shape {
            appearance PBRAppearance {
              emissiveColor 0 1 0
              transparency 0.3
              metalness 0
              roughness 0.7
              baseColorMap ImageTexture {
                url [
                  "https://raw.githubusercontent.com/cyberbotics/webots/R2022b/projects/robots/mobsya/thymio/protos/textures/thymio2_battery_level.png"
                ]
              }
            }
            geometry Plane {
              size 0.006 0.01
            }
          }
        ]
      }
      HingeJoint {
        jointParameters HingeJointParameters {
          anchor 0 -0.05 0.021
          axis 0 1 0
        }
        device [
          RotationalMotor {
            name "motor.right"
            maxVelocity 9.53
          }
          PositionSensor {
            name "motor.right.position"
          }
        ]
        endPoint Solid {
          translation 0 -0.05 0.021
          rotation 0.57735 0.57735 0.57735 2.094395
          children [
            DEF WHEEL_SHAPE Shape {
              appearance DEF WHITE_PLASTIC_APPEARANCE PBRAppearance {
                baseColor 0.9 0.9 0.9
                roughness 0.1
                metalness 0
              }
              geometry Thymio2WheelIndexedFaceSet {
              }
            }
            DEF TIRE_SHAPE Shape {
              appearance PBRAppearance {
                baseColor 0.1 0.1 0.1
                metalness 0
                roughness 0.3
              }
              geometry Thymio2TireIndexedFaceSet {
              }
            }
          ]
          name "right wheel"
          boundingObject Transform {
            rotation -0.577349 -0.577351 0.577351 2.094397
            children [
              DEF WHEEL_BO_SHAPE Shape {
                geometry Cylinder {
                  height 0.015
                  radius 0.021
                }
              }
            ]
          }
          physics Physics {
            density -1
            mass 0.024
          }
        }
      }
      HingeJoint {
        jointParameters HingeJointParameters {
          anchor 0 0.05 0.021
          axis 0 1 0
        }
        device [
          RotationalMotor {
            name "motor.left"
            maxVelocity 9.53
          }
          PositionSensor {
            name "motor.left.position"
          }
        ]
        endPoint Solid {
          translation 0 0.05 0.021
          rotation 0.57735 0.57735 0.57735 2.094395
          children [
            Transform {
              rotation 0 1 0 3.14159
              children [
                USE WHEEL_SHAPE
                USE TIRE_SHAPE
              ]
            }
          ]
          name "left wheel"
          boundingObject Transform {
            rotation -0.577349 -0.577351 0.577351 2.094397
            children [
              DEF WHEEL_BO_SHAPE Shape {
                geometry Cylinder {
                  height 0.015
                  radius 0.021
                }
              }
            ]
          }
          physics Physics {
            density -1
            mass 0.024
          }
        }
      }
      TouchSensor {
        translation 0.0489 0.015 0.0526
        rotation 0 0 1 1.570796
        name "button.left"
        children [
          DEF TOUCH_SENSOR_TRANSFORM Transform {
            children [
              Shape {
                appearance USE MAIN_APPEARANCE
                geometry Plane {
                  size 0.015 0.015
                }
              }
            ]
          }
        ]
      }
      TouchSensor {
        translation 0.0639 0 0.0526
        rotation 0 0 1 1.570796
        name "button.forward"
        children [
          USE TOUCH_SENSOR_TRANSFORM
        ]
      }
      TouchSensor {
        translation 0.0489 -0.015 0.0526
        rotation 0 0 1 1.570796
        name "button.right"
        children [
          USE TOUCH_SENSOR_TRANSFORM
        ]
      }
      TouchSensor {
        translation 0.0339 0 0.0526
        rotation 0 0 1 1.570796
        name "button.backward"
        children [
          USE TOUCH_SENSOR_TRANSFORM
        ]
      }
      TouchSensor {
        translation 0.0489 0 0.0526
        rotation 0 0 1 1.570796
        name "button.center"
        children [
          USE TOUCH_SENSOR_TRANSFORM
        ]
      }
      
      Thymio2DistanceSensor {
        translation 0.05 -0.06 0.04
        rotation 0 0 -1 1.5708
        name "prox.thymio"
      }
      
      DistanceSensor {
        rotationStep 0.05
        translation 0.05 -0.06 0.03
        rotation 0 0 -1 1.5708
        name "prox.laser"
        type "laser"
        aperture 0.001
        lookupTable [
          0.0 4235 0.01
          0.00333 4308 0.01
          0.00666 4298 0.01
          0.01 4279 0.01
          0.01333 4253 0.01
          0.01666 4201 0.01
          0.02 4139 0.01
          0.02333 3981 0.01
          0.02666 3733 0.01
          0.03 3515 0.01
          0.03333 3326 0.01
          0.03666 3141 0.01
          0.04 2965 0.01
          0.04333 2860 0.01
          0.04666 2696 0.01
          0.05 2570 0.01
          0.05333 2475 0.01
          0.05666 2344 0.01
          0.06 2248 0.01
          0.06333 2182 0.01
          0.06666 2044 0.01
          0.07 1941 0.01
          0.07333 1835 0.01
          0.07666 1689 0.01
          0.08 1597 0.01
          0.08333 1473 0.01
          0.08666 1345 0.01
          0.09 1195 0.01
          0.09333 0 0.01
        ]
        children [
          Shape {
            appearance PBRAppearance {
              baseColor 0 0 0
              metalness 0
              roughness 0.4
            }
            geometry IndexedFaceSet {
              coord Coordinate {
                point [
                  -0.00056 -0.000875 -0.001247 -0.002013 -0.001744 -0.002141 -0.002011 -0.001232 -0.003391 -0.000559 -0.001233 -0.002136 -0.002014 -0.001237 -0.000883 -0.000561 -4e-06 -0.000874 -0.00056 0.00123 -0.002117 4.3e-05 -2e-06 -0.002126 -0.000561 0.000867 -0.001235 -0.002011 1e-06 -0.0039 -0.002012 0.001233 -0.003373 -0.000558 -0.000871 -0.003019 -0.000559 0.000871 -0.003007 -0.000558 1e-06 -0.00338 -0.002014 -5e-06 -0.000356 -0.002014 0.001228 -0.000865 -0.002013 0.00174 -0.002115 0.004797 0.117094 0.021731 -1e-06 -0.001786 -0.004009 -6e-06 -0.001795 0.00025 -4e-06 -0.001794 -0.000268 -2e-06 -0.002296 -0.004533 -0.001125 -0.001786 -0.004011 -2e-06 0.002294 -0.004502 -2e-06 0.001789 -0.003984 -5e-06 0.001782 -0.000242 -6e-06 0.001781 0.000277 -0.002018 0.002279 0.004502 -0.00201 -0.002297 -0.004538 -0.00203 -0.001762 -0.004014 -0.002031 0.001789 -0.003988 -9e-06 -0.002314 0.004508 -9e-06 -0.001802 0.003991 -0.002037 -0.001777 0.003977 -0.002034 -0.00177 0.000246 -0.002034 -0.001768 -0.000284 -0.002018 -0.002315 0.004502 -9e-06 0.002284 0.004542 -0.00201 0.002297 -0.004538 -0.001132 0.001773 0.004007 -0.001133 0.001774 0.004017 -9e-06 0.001774 0.004018 -0.002034 0.001781 -0.000257 -0.002035 0.001781 0.000273 -0.002037 0.001773 0.004004
                ]
              }
              coordIndex [
                2, 3, 1, -1, 0, 4, 3, -1, 4, 1 3, -1, 5, 7, 8, -1, 7, 5, 0, -1 7, 6, 8, -1, 0, 3, 7, -1, 3, 11 7, -1, 2, 11, 3, -1, 6, 7, 12, -1 10, 6, 12, -1, 13, 12, 7, -1, 11, 13 7, -1, 9, 10, 12, -1, 13, 9, 12, -1 2, 9, 13, -1, 11, 2, 13, -1, 5, 4 0, -1, 14, 5, 8, -1, 14, 4, 5, -1 14, 8, 15, -1, 16, 6, 10, -1, 16, 15 8, -1, 6, 16, 8, -1, 18, 24, 30, -1 19, 32, 31, -1, 21, 18, 20, -1, 30, 29 18, -1, 20, 18, 29, -1, 28, 23, 21, -1 37, 26, 23, -1, 26, 25, 23, -1, 28, 38 23, -1, 24, 23, 25, -1, 30, 24, 25, -1 25, 19, 20, -1, 20, 19, 31, -1, 21, 20 31, -1, 31, 32, 41, -1, 35, 42, 20, -1 31, 36, 21, -1, 28, 21, 36, -1, 44, 41 33, -1, 32, 40, 41, -1, 41, 37, 31, -1 20, 29, 35, -1, 37, 36, 31, -1, 41, 39 26, -1, 26, 37, 41, -1, 25, 26, 19, -1 41, 32, 33, -1, 20, 42, 25, -1, 32, 19 33, -1, 26, 39, 43, -1, 25, 42, 30, -1 43, 39, 44, -1, 33, 19, 34, -1, 34, 19 43, -1, 26, 43, 19, -1, 18, 23, 24, -1 21, 23, 18, -1, 27, 36, 37, -1, 27, 37 38, -1, 23, 38, 37, -1, 38, 27, 28, -1 28, 27, 36, -1
              ]
              creaseAngle 1
            }
          }
          Shape {
            appearance PBRAppearance {
              baseColor 0.392157 0.392157 0.392157
              roughness 0
              metalness 0
            }
            geometry IndexedFaceSet {
              coord Coordinate {
                point [
                  -0.001984 -0.001817 0.002114 -0.001983 -0.001283 0.000812 -0.000513 -0.001286 0.002119 -0.000513 -0.000911 0.003045 -0.000511 -0.000908 0.001199 -0.001985 -6e-06 0.003973 -0.001985 -0.001288 0.003423 -0.001981 2e-06 0.000281 9.7e-05 -2e-06 0.002129 -0.000512 0.000908 0.001212 -0.000514 0.000904 0.003058 -0.000511 0 0.000823 -0.001984 0.001813 0.002141 -0.001982 0.001284 0.000832 -0.000513 0.001282 0.002138 -0.001985 0.001279 0.003442 -0.000514 -4e-06 0.003434
                ]
              }
              coordIndex [
                0, 2, 6, -1, 2, 4, 8, -1, 8, 3 2, -1, 2, 0, 1, -1, 2, 1, 4, -1 5, 6, 16, -1, 16, 6, 3, -1, 6, 2 3, -1, 16, 15, 5, -1, 16, 3, 8, -1 8, 9, 14, -1, 14, 10, 8, -1, 10, 16 8, -1, 8, 11, 9, -1, 8, 4, 11, -1 1, 7, 11, -1, 7, 9, 11, -1, 4, 1 11, -1, 14, 12, 10, -1, 12, 9, 13, -1 12, 14, 9, -1, 9, 7, 13, -1, 15, 16 10, -1, 12, 15, 10, -1
              ]
              creaseAngle 1
            }
          }
        ]
      }
      
      Thymio2LED {
        translations [ 0.0582 0 0.0532 ]
        rotations [ 0 0 1 1.570796 ]
        name "leds.buttons.led0"
        textureUrls [ "https://raw.githubusercontent.com/cyberbotics/webots/R2022b/projects/robots/mobsya/thymio/protos/textures/thymio2_square_flare.png" ]
        sizes [ 0.01 0.003 ]
      }
      Thymio2LED {
        translations [ 0.0486 0 0.0532 ]
        rotations [ 0 0 1 -3.141589 ]
        name "leds.buttons.led1"
        textureUrls [ "https://raw.githubusercontent.com/cyberbotics/webots/R2022b/projects/robots/mobsya/thymio/protos/textures/thymio2_square_flare.png" ]
        sizes [ 0.01 0.003 ]
      }
      Thymio2LED {
        translations [ 0.0398 0 0.0532 ]
        rotations [ 0 0 1 1.570796 ]
        name "leds.buttons.led2"
        textureUrls [ "https://raw.githubusercontent.com/cyberbotics/webots/R2022b/projects/robots/mobsya/thymio/protos/textures/thymio2_square_flare.png" ]
        sizes [ 0.01 0.003 ]
      }
      Thymio2LED {
        translations [ 0.0486 0 0.0532 ]
        rotations [ 0 0 1 -3.141589 ]
        name "leds.buttons.led3"
        textureUrls [ "https://raw.githubusercontent.com/cyberbotics/webots/R2022b/projects/robots/mobsya/thymio/protos/textures/thymio2_square_flare.png" ]
        sizes [ 0.01 0.003 ]
      }
      Thymio2LED {
        translations [ 0.0732 0 0.0532 ]
        rotations [ 0 0 1 0 ]
        name "leds.circle.led0"
        textureUrls [ "https://raw.githubusercontent.com/cyberbotics/webots/R2022b/projects/robots/mobsya/thymio/protos/textures/thymio2_curved_flare.png" ]
        sizes [ 0.004 0.016 ]
        color 0.5 0.25 0.125
      }
      Thymio2LED {
        translations [ 0.066 -0.0172 0.0532 ]
        rotations [ 0 0 1 -0.785398 ]
        name "leds.circle.led1"
        textureUrls [ "https://raw.githubusercontent.com/cyberbotics/webots/R2022b/projects/robots/mobsya/thymio/protos/textures/thymio2_curved_flare.png" ]
        sizes [ 0.004 0.016 ]
        color 0.5 0.25 0.125
      }
      Thymio2LED {
        translations [ 0.0489 -0.0243 0.0532 ]
        rotations [ 0 0 1 -1.570796 ]
        name "leds.circle.led2"
        textureUrls [ "https://raw.githubusercontent.com/cyberbotics/webots/R2022b/projects/robots/mobsya/thymio/protos/textures/thymio2_curved_flare.png" ]
        sizes [ 0.004 0.016 ]
        color 0.5 0.25 0.125
      }
      Thymio2LED {
        translations [ 0.0317 -0.0172 0.0532 ]
        rotations [ 0 0 1 -2.356194 ]
        name "leds.circle.led3"
        textureUrls [ "https://raw.githubusercontent.com/cyberbotics/webots/R2022b/projects/robots/mobsya/thymio/protos/textures/thymio2_curved_flare.png" ]
        sizes [ 0.004 0.016 ]
        color 0.5 0.25 0.125
      }
      Thymio2LED {
        translations [ 0.0246 0 0.0532 ]
        rotations [ 0 0 1 -3.141593 ]
        name "leds.circle.led4"
        textureUrls [ "https://raw.githubusercontent.com/cyberbotics/webots/R2022b/projects/robots/mobsya/thymio/protos/textures/thymio2_curved_flare.png" ]
        sizes [ 0.004 0.016 ]
        color 0.5 0.25 0.125
      }
      Thymio2LED {
        translations [ 0.0317 0.0172 0.0532 ]
        rotations [ 0 0 1 2.356194 ]
        name "leds.circle.led5"
        textureUrls [ "https://raw.githubusercontent.com/cyberbotics/webots/R2022b/projects/robots/mobsya/thymio/protos/textures/thymio2_curved_flare.png" ]
        sizes [ 0.004 0.016 ]
        color 0.5 0.25 0.125
      }
      Thymio2LED {
        translations [ 0.0489 0.0243 0.0532 ]
        rotations [ 0 0 1 1.570796 ]
        name "leds.circle.led6"
        textureUrls [ "https://raw.githubusercontent.com/cyberbotics/webots/R2022b/projects/robots/mobsya/thymio/protos/textures/thymio2_curved_flare.png" ]
        sizes [ 0.004 0.016 ]
        color 0.5 0.25 0.125
      }
      Thymio2LED {
        translations [ 0.066 0.0172 0.0532 ]
        rotations [ 0 0 1 0.785398 ]
        name "leds.circle.led7"
        textureUrls [ "https://raw.githubusercontent.com/cyberbotics/webots/R2022b/projects/robots/mobsya/thymio/protos/textures/thymio2_curved_flare.png" ]
        sizes [ 0.004 0.016 ]
        color 0.5 0.25 0.125
      }
      Thymio2LED {
        name "leds.prox.h.led0"
        translations [ 0.0589 0.0556 0.0322 ]
        rotations [ 0.257494 0.683264 0.683262 2.637561 ]
        textureUrls [ "https://raw.githubusercontent.com/cyberbotics/webots/R2022b/projects/robots/mobsya/thymio/protos/textures/thymio2_radial_flare.png" ]
        sizes [ 0.01 0.01 ]
      }
      Thymio2LED {
        name "leds.prox.h.led1"
        translations [ 0.0733 0.0354 0.0322 ]
        rotations [ 0.405581 0.646339 0.646336 2.370983 ]
        textureUrls [ "https://raw.githubusercontent.com/cyberbotics/webots/R2022b/projects/robots/mobsya/thymio/protos/textures/thymio2_radial_flare.png" ]
        sizes [ 0.01 0.01 ]
      }
      Thymio2LED {
        name "leds.prox.h.led2"
        translations [ 0.0809 0 0.0322 ]
        rotations [ 0.51968 0.604124 0.604125 2.183057 ]
        textureUrls [ "https://raw.githubusercontent.com/cyberbotics/webots/R2022b/projects/robots/mobsya/thymio/protos/textures/thymio2_radial_flare.png" ]
        sizes [ 0.01 0.01 ]
      }
      Thymio2LED {
        name "leds.prox.h.led3"
        translations [ 0.0809 0 0.0322 ]
        rotations [ 0.635005 0.546245 0.546246 2.010068 ]
        textureUrls [ "https://raw.githubusercontent.com/cyberbotics/webots/R2022b/projects/robots/mobsya/thymio/protos/textures/thymio2_radial_flare.png" ]
        sizes [ 0.01 0.01 ]
      }
      Thymio2LED {
        name "leds.prox.h.led4"
        translations [ 0.0733 -0.0354 0.0322 ]
        rotations [ -0.747954 -0.469345 -0.46934 -1.857217 ]
        textureUrls [ "https://raw.githubusercontent.com/cyberbotics/webots/R2022b/projects/robots/mobsya/thymio/protos/textures/thymio2_radial_flare.png" ]
        sizes [ 0.01 0.01 ]
      }
      Thymio2LED {
        name "leds.prox.h.led5"
        translations [ 0.059 -0.0557 0.0322 ]
        rotations [ 0.882491 0.332575 0.33257 1.695481 ]
        textureUrls [ "https://raw.githubusercontent.com/cyberbotics/webots/R2022b/projects/robots/mobsya/thymio/protos/textures/thymio2_radial_flare.png" ]
        sizes [ 0.01 0.01 ]
      }
      Thymio2LED {
        name "leds.prox.h.led6"
        translations [ -0.0308 0.0406 0.0322 ]
        rotations [ -0.57735 -0.57735 0.577351 2.094395 ]
        textureUrls [ "https://raw.githubusercontent.com/cyberbotics/webots/R2022b/projects/robots/mobsya/thymio/protos/textures/thymio2_radial_flare.png" ]
        sizes [ 0.01 0.01 ]
      }
      Thymio2LED {
        name "leds.prox.h.led7"
        translations [ -0.0308 -0.0406 0.0322 ]
        rotations [ -0.57735 -0.57735 0.577351 2.094395 ]
        textureUrls [ "https://raw.githubusercontent.com/cyberbotics/webots/R2022b/projects/robots/mobsya/thymio/protos/textures/thymio2_radial_flare.png" ]
        sizes [ 0.01 0.01 ]
      }
      Thymio2LED {
        name "leds.sound"
        translations [ 0.0267 -0.0587 0.0367 ]
        rotations [ 0.577351 -0.577351 0.577349 2.094397 ]
        textureUrls [ "https://raw.githubusercontent.com/cyberbotics/webots/R2022b/projects/robots/mobsya/thymio/protos/textures/thymio2_radial_flare.png" ]
        sizes [ 0.01 0.01 ]
        color 0 0 1
      }
      Thymio2LED {
        name "leds.rc"
        translations [ 0.0417 -0.0587 0.0367 ]
        rotations [ 0.577351 -0.577351 0.577349 2.094397 ]
        textureUrls [ "https://raw.githubusercontent.com/cyberbotics/webots/R2022b/projects/robots/mobsya/thymio/protos/textures/thymio2_radial_flare.png" ]
        sizes [ 0.01 0.01 ]
      }
      Thymio2LED {
        name "leds.temperature.red"
        translations [ 0.036 0.0587 0.0367 ]
        rotations [ -0.577351 0.577351 0.577349 2.094397 ]
        textureUrls [ "https://raw.githubusercontent.com/cyberbotics/webots/R2022b/projects/robots/mobsya/thymio/protos/textures/thymio2_radial_flare.png" ]
        sizes [ 0.01 0.01 ]
      }
      Thymio2LED {
        name "leds.temperature.blue"
        translations [ 0.029 0.0587 0.0367 ]
        rotations [ -0.577351 0.577351 0.577349 2.094397 ]
        textureUrls [ "https://raw.githubusercontent.com/cyberbotics/webots/R2022b/projects/robots/mobsya/thymio/protos/textures/thymio2_radial_flare.png" ]
        sizes [ 0.01 0.01 ]
        color 0 0 1
      }
      Thymio2LED {
        name "leds.prox.v.led0"
        translations [ 0.0733 0.0222 0.00790001, 0.0778 0.0231 0.0156]
        rotations [ 0.707107 0.707107 0 3.141527, 0.476903 0.621515 0.621517 2.251588]
        textureUrls [ "https://raw.githubusercontent.com/cyberbotics/webots/R2022b/projects/robots/mobsya/thymio/protos/textures/thymio2_radial_flare.png", "textures/thymio2_radial_flare.png"]
        sizes [ 0.01 0.01, 0.01 0.01 ]
        %< if (fields.castLight.value) { >%
          lightPosition 0.073 0.022 0.03
        %< } >%
      }
      Thymio2LED {
        name "leds.prox.v.led1"
        translations [ 0.0733 -0.0222 0.00790001, 0.0778 -0.0231 0.0156]
        rotations [ 0.707107 0.707107 0 3.141527, -0.67766 -0.519986 -0.519992 -1.950437]
        textureUrls [ "https://raw.githubusercontent.com/cyberbotics/webots/R2022b/projects/robots/mobsya/thymio/protos/textures/thymio2_radial_flare.png", "textures/thymio2_radial_flare.png"]
        sizes [ 0.01 0.01, 0.01 0.01 ]
        %< if (fields.castLight.value) { >%
          lightPosition 0.073 -0.022 0.03
        %< } >%
      }
      Thymio2LED {
        name "leds.top"
        translations [
          0 0.0319 0.0532,
          -0.0308 0.0318 0.045,
          0 0.0398 0.041,
          0 -0.0319 0.0532,
          -0.0308 -0.0318 0.045,
          0 -0.0398 0.041
        ]
        rotations [
          0 0 1 -1.570889,
          -0.57735 -0.57735 0.577351 2.094395,
          -1 0 0 1.570799,
          0 0 1 -1.570889,
          -0.57735 -0.57735 0.577351 2.094395,
          0 -0.707108 0.707106 -3.141589
        ]
        textureUrls [
          "https://raw.githubusercontent.com/cyberbotics/webots/R2022b/projects/robots/mobsya/thymio/protos/textures/thymio2_radial_flare.png",
          "https://raw.githubusercontent.com/cyberbotics/webots/R2022b/projects/robots/mobsya/thymio/protos/textures/thymio2_square_flare.png",
          "https://raw.githubusercontent.com/cyberbotics/webots/R2022b/projects/robots/mobsya/thymio/protos/textures/thymio2_radial_flare.png",
          "https://raw.githubusercontent.com/cyberbotics/webots/R2022b/projects/robots/mobsya/thymio/protos/textures/thymio2_radial_flare.png",
          "https://raw.githubusercontent.com/cyberbotics/webots/R2022b/projects/robots/mobsya/thymio/protos/textures/thymio2_square_flare.png",
          "https://raw.githubusercontent.com/cyberbotics/webots/R2022b/projects/robots/mobsya/thymio/protos/textures/thymio2_radial_flare.png",
        ]
        sizes [
          0.05 0.06,
          0.04 0.014,
          0.05 0.02,
          0.05 0.06,
          0.04 0.014,
          0.05 0.02,
        ]
        level 0
      }
      Thymio2LED {
        name "leds.bottom.left"
        translations [ 0.0459 -0.0582 0.0213, 0.0626 -0.0513 0.0213]
        rotations [ 1 0 0 1.570799, 0.692566 -0.244255 0.678743 2.671006]
        textureUrls [ "https://raw.githubusercontent.com/cyberbotics/webots/R2022b/projects/robots/mobsya/thymio/protos/textures/thymio2_radial_flare.png", "textures/thymio2_radial_flare.png"]
        sizes [ 0.02 0.02, 0.02 0.01 ]
        level 0
        %< if (fields.castLight.value) { >%
          lightPosition 0.05 -0.04 0.03
        %< } >%
      }
      Thymio2LED {
        name "leds.bottom.right"
        translations [ 0.0459 0.0582 0.0213, 0.0627 0.0514 0.0213]
        rotations [ 0 0.707107 0.707107 -3.141501, 0.688836 0.248277 0.681075 -2.631613]
        textureUrls [ "https://raw.githubusercontent.com/cyberbotics/webots/R2022b/projects/robots/mobsya/thymio/protos/textures/thymio2_radial_flare.png", "textures/thymio2_radial_flare.png"]
        sizes [ 0.02 0.02, 0.02 0.01 ]
        level 0
        %< if (fields.castLight.value) { >%
          lightPosition 0.05 0.04 0.03
        %< } >%
      }
      Accelerometer {
        name "acc"
        translation 0.015 -0.02 0.03
        rotation 0 0 1 -1.570798
        lookupTable [
          -13.65 -32 0
          13.65 32 0
        ]
      }
      Group {
        children IS bodySlot
      }
    ]
    contactMaterial IS contactMaterial
    boundingObject Group {
      children [
        Transform {
          translation 0.013 0 0.0316
          rotation 0.57735 0.57735 0.57735 2.094395
          children [
            Box {
              size 0.115 0.0442 0.085
            }
          ]
        }
        Transform {
          translation 0.0677 0 0.0316
          rotation 0.57735 0.57735 0.57735 2.094395
          children [
             DEF FRONT_BOX Box {
              size 0.0245 0.0442 0.024
            }
          ]
        }
        Transform {
          translation 0.0643 0.0214 0.0316
          rotation -0.457027 -0.628938 -0.628938 -2.284229
          children [
            USE FRONT_BOX
          ]
        }
        Transform {
          translation 0.0542 0.0403 0.0316
          rotation -0.330927 -0.667266 -0.667266 -2.502426
          children [
            USE FRONT_BOX
          ]
        }
        Transform {
          translation 0.0643 -0.0214 0.0316
          rotation -0.697397 -0.506773 -0.506773 -1.92364
          children [
            USE FRONT_BOX
          ]
        }
        Transform {
          translation 0.0542 -0.0403 0.0316
          rotation 0.818704 0.406032 0.406032 1.769508
          children [
            USE FRONT_BOX
          ]
        }
        Transform {
          translation 0.0612 0 0.009
          rotation 0.57735 0.57735 0.57735 2.094395
          children [
            Sphere {
              radius 0.0096
            }
          ]
        }
      ]
    }
    physics Physics {
      density -1
      mass 0.2
    }
  }
}
